### ISSUE TITLE
returning multiple subnets in data.aws_subnet.subnet.*.id results in Terraform Crash

### ISSUE THREAD

**Issue by icornett** [hashicorp/terraform#16902]
### Terraform Version
0.11.1

```
...
```

### Terraform Configuration Files

```hcl
data "aws_subnet" "frontend" {
  id = "${aws_subnet.subnet.*.id}"
}

data "aws_subnet" "backend" {
  id = "${aws_subnet.be_subnet.*.id}"
}

# Create main subnet for network - balance
resource "aws_subnet" "subnet" {
  count             = "${length(data.aws_availability_zones.all.names)}"
  vpc_id            = "${data.aws_vpc.selected.id}"
  availability_zone = "${element(data.aws_availability_zones.all.names, count.index)}"
  cidr_block        = "${cidrsubnet(data.aws_vpc.selected.cidr_block, 8, count.index + 1)}"

  #  ipv6_cidr_block   = "${data.aws_vpc.selected.ipv6_cidr_block}"

  lifecycle {
    #    prevent_destroy = true
  }
  tags {
    Name = "upwork-demo-public-subnet-${var.environmentName}-${count.index}"
  }
}

# Create private network for BE
resource "aws_subnet" "be_subnet" {
  count             = "${length(data.aws_availability_zones.all.names)}"
  vpc_id            = "${data.aws_vpc.selected.id}"
  availability_zone = "${element(data.aws_availability_zones.all.names, count.index)}"
  cidr_block        = "${cidrsubnet(data.aws_vpc.selected.cidr_block, 8, count.index + 100)}"

  #  ipv6_cidr_block   = "${data.aws_vpc.selected.ipv6_cidr_block}"

  lifecycle {
    #    prevent_destroy = true
  }
  tags {
    Name = "upwork-demo-private-subnet-${var.environmentName}-${count.index}"
  }
}
```

### Crash Output

https://gist.github.com/icornett/de17f71bffd496dfe018a7da5b0f827e

### Expected Behavior

return list of subnets

### Actual Behavior

App crashes

### Steps to Reproduce
terraform init
terraform plan
terraform apply

**Comment by jgrossmac** [comment]
I have this same issue.  Was working great before.

**Comment by icornett** [comment]
It looks like you're forgetting a variable name (presumably in vars.tf) named environmentName

e.g. 
```
variable "environmentName" {
  description = "The environment name, i.e. prod"
  default        = "dev"
}
```

with regard to the unknown resource errors, it looks like you may not be calling the data sources correctly.   I think you'd want to do `data.aws_availability_zones.all.*.names` instead of `data.aws_availability_zones.all.names`

**Comment by igaskin** [comment]
I had the same issue, and used this as my work-around.  

```
data "aws_vpc" "default" {
  state = "available"

  tags {
    Name = "${var.vpc_name}"
  }
}

data "aws_subnet_ids" "private" {
  vpc_id = "${data.aws_vpc.default.id}"

  tags {
    Name = "west-private-*"
  }
}
```

Tested with terraform v0.11.7

**Comment by apparentlymart** [comment]
Hi all,

The `id` argument for this data source expects a single string rather than a list, so the given configuration isn't correct but it is definitely a bug that the provider SDK crashed rather than returning a proper error message for it.

In the next major release Terraform's architecture will be changing so that the responsibility for validating this particular problem will belong to Terraform Core rather than to the provider SDK, and so I've re-labelled this to put it in the scope of the testing pass we'll do in preparation for that release. At that point, the behavior we'd expect here is an explicit error message saying that a string value is required for this argument, rather than a crash.

As @igaskin noted, if you wish to retrieve several subnets at once you can use [`aws_subnet_ids`](https://www.terraform.io/docs/providers/aws/d/subnet_ids.html) (plural) to find the set of subnet ids that match a given set of criteria, or if you already know a specific set of ids you wish to retrieve then you can use `count` with `aws_subnet` to retrieve each of them as a separate instance of that data source.

**Comment by apparentlymart** [comment]
Hi again!

The configuration language improvement work has now landed in master, so I was able to verify this with the v0.12.0-alpha1 build using the following configuration that I adapted from the one in the opening comment on this issue:

```hcl
resource "null_resource" "example" {
  count = 2
}

data "aws_subnet" "frontend" {
  id = null_resource.example.*.id
}
```

This configuration now produces a proper error, rather than a crash:

```
$ terraform apply
Error: Incorrect attribute value type

  on data-id-list-crash.tf line 6, in data "aws_subnet" "frontend":
   6:   id = null_resource.example.*.id

Inappropriate value for attribute "id": string required.
```

Since the crash is now fixed in master, I'm going to close this out. The fix will be included in the forthcoming v0.12.0 final release. Thanks for reporting this, and thanks for your patience while we got this fixed up.

**Comment by sonianara** [comment]
@apparentlymart 
```
resource "null_resource" "example" {
  count = 2
}

data "aws_subnet" "frontend" {
  id = null_resource.example.*.id
}
```
What would aws_subnet.frontend look like in v12?
